### PR TITLE
Use torch.compile for version 2.0 and higher

### PR DESCRIPTION
**Pull request by ksivaman**

### CODE REVIEW

File: transformer_engine/pytorch/fp8.py [hunk 1]
```diff
@@ -35,6 +37,10 @@
 _dp_amax_reduce_forward_idx = 0
 _dp_amax_reduce_backward_idx = 0
 
+jit_fuser = torch.jit.script
+if packaging.version.Version(version("torch")) >= packaging.version.Version("2"):
```
Comment (ptrendx) [inline transformer_engine/pytorch/fp8.py:41]: Why not use something like `torch.__version__`?

**Comment by ptrendx** [comment]
Any insight on performance difference between the 2 approaches to the various kernels?

File: transformer_engine/pytorch/fp8.py (hunk 1 above)
Comment (ksivaman) [reply transformer_engine/pytorch/fp8.py:41]: Because `torch.__version__` is of specific type `TorchVersion` which makes comparisons non-trivial to do.

File: transformer_engine/pytorch/fp8.py [hunk 2]
```diff
@@ -35,6 +37,10 @@
...
+    jit_fuser = torch.compile
```
Comment (timmoon10) [inline transformer_engine/pytorch/fp8.py:42]:
```suggestion
from .jit import jit_fuser
```

File: transformer_engine/pytorch/fp8.py (hunk 1 above)
Comment (timmoon10) [reply transformer_engine/pytorch/fp8.py]: From what I can tell, `TorchVersion` has a pretty convenient interface:
```python
if torch.__version__ >= "2":
```
See: https://github.com/pytorch/pytorch/blob/32360b48e8d09984beb831ccb4a9d91b2de47189/torch/torch_version.py#L41

File: transformer_engine/pytorch/fp8.py [hunk 3]
```diff
@@ -4,16 +4,19 @@
 
 """FP8 utilities for TransformerEngine"""
 import os
+from importlib.metadata import version
```
Comment (timmoon10) [inline transformer_engine/pytorch/fp8.py:7]:
```suggestion
```

File: transformer_engine/pytorch/fp8.py [hunk 4]
```diff
@@ -4,16 +4,19 @@
...
 from contextlib import contextmanager
 from collections import deque
 from typing import Callable, List, Optional, Dict, Any, Tuple, Union
+from pkg_resources import packaging
```
Comment (timmoon10) [inline transformer_engine/pytorch/fp8.py:11]:
```suggestion
```

**Review by timmoon10 — APPROVED** [review]
LGTM

**Comment by ksivaman** [comment]
For the dropout fusion cases (training and inference), NVfuser is better for perf than `torch.compile` since `torch.compile` generates 3 kernels instead of one, including a auxiliary distribution kernel and a memcpy kernel, while also being much slower and incurring non-negligible CPU overhead. For all other cases `torch.compile` is equal or better, i.e. generates lesser number of and/or faster kernels.

**Review by timmoon10 — APPROVED** [review]
LGTM. It would be helpful to add a comment to the dropout fusion operations explaining why they're using TorchScript instead of `torch.compile`.

**Review by ptrendx — APPROVED** [review]

**Comment by ksivaman** [comment]
Pipeline 8562248
